hydra-node/test/Hydra/HeadLogicSpec.hs

```diff
@@ -455,27 +455,33 @@ observationEvent observedTx =
     }
 
 hasEffect :: (HasCallStack, IsChainState tx) => Outcome tx -> Effect tx -> IO ()
-hasEffect (NewState _ effects) effect
-  | effect `elem` effects = pure ()
-  | otherwise = failure $ "Missing effect " <> show effect <> " in produced effects: " <> show effects
-hasEffect (OnlyEffects effects) effect
-  | effect `elem` effects = pure ()
-  | otherwise = failure $ "Missing effect " <> show effect <> " in produced effects: " <> show effects
-hasEffect o _ = failure $ "Unexpected outcome: " <> show o
+hasEffect outcome effect =
+  case outcome of
+    Effects effects
+      | effect `elem` effects -> pure ()
+      | otherwise -> failure $ "Missing effect " <> show effect <> " in produced effects: " <> show effects
+    Combined l r ->
+      hasEffect l effect `orElse` hasEffect r effect
+    _ -> failure $ "Unexpected outcome: " <> show outcome
```

Comment:
Having the capability to collect all the effects here could make a better code here.